File: tcp-ip/ch15/go/io/io.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"syscall"
	"time"
)
const BUF_SIZE int = 3
func sysIo() {
	start := time.Now()
	readFd, _ := syscall.Open("../../README.md", os.O_RDONLY, 0755)
    //! O_CREAT 文件不存在则创建文件
	//! O_RDONLY 只读、O_WRONLY 只写、O_TRUNC 重写
	writeFd, _ := syscall.Open("../../README.txt", os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0755)
	buf := make([]byte, BUF_SIZE)
	for {
		nBytes, _ := syscall.Read(readFd, buf)
		if nBytes == 0 {
			break
		}
		syscall.Write(writeFd, buf[:nBytes])
	}
	syscall.Close(readFd)
	syscall.Close(writeFd)
	total := time.Since(start)
	fmt.Println("Sys IO uses", total)
}

func stdIo() {
	start := time.Now()
	readFp, _ := os.OpenFile("../../README.md", os.O_RDONLY, 0755)
	writeFp, _ := os.OpenFile("../../README.txt", os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0755)
	reader := bufio.NewReader(readFp)
	writer := bufio.NewWriter(writeFp)
	buf := make([]byte, BUF_SIZE)
	for {
		nBytes, _ := reader.Read(buf)
		if nBytes == 0 {
			break
		}
		writer.Write(buf[:nBytes])
	}
	readFp.Close()
	writeFp.Close()
	total := time.Since(start)
	fmt.Println("Std IO total", total)
}

func main() {
	sysIo()
	stdIo()
}

File: tcp-ip/ch15/io.cc
//
// Created by Tiancheng on 2024/8/29.
//

#include "ctime"
#include <cstdio>
#include <fcntl.h>
#include <unistd.h>

#define BUF_SIZE 3

void sysIo() {
    clock_t start = clock();
    int readFd = open("../README.md", O_RDONLY);
    int writeFd = open("../README.txt", O_WRONLY | O_CREAT | O_TRUNC);
    char buf[BUF_SIZE];
    while (true) {
        int readBytes = read(readFd, buf, BUF_SIZE);
        if (readBytes <= 0) {
            break;
        }
        write(writeFd, buf, readBytes);
    }
    close(readFd);
    close(writeFd);
    clock_t end = clock();
    printf("Sys IO costs %fms\n", ((double)(end - start)) / 1000);
}

void stdIo() {
    clock_t start = clock();
    FILE *readFd = fopen("../README.md", "r");
    FILE *writeFd = fopen("../README.txt", "w");
    char buf[BUF_SIZE];
    while (true) {
        char *row = fgets(buf, BUF_SIZE, readFd);
        if (row == NULL) {
            break;
        }
        fputs(buf, writeFd);
    }
    fclose(readFd);
    fclose(writeFd);
    clock_t end = clock();
    printf("Std IO total %fms\n", ((double)(end - start)) / 1000);
}

int main() {
    sysIo();
    stdIo();
}